django/db/migrations/autodetector.py

```diff
@@ -610,16 +610,16 @@ def generate_moved_models(self):
                 ):
                     continue
                 dependencies = []
-                # Quit if db_table is not explicitly provided in moved model.
-                if not model_state.options.get(
-                    "db_table",
-                ):
-                    db_table = f"{app_label}_{model_name}"
+                old_table_name = rem_model_state.options.get("db_table")
+                new_table_name = model_state.options.get("db_table")
+                # If moved model has same table name as it was in old app,
+                # then don't rename table.
+                if old_table_name == new_table_name:
```

Comment:
1. The logic is reversed, the rename should happen if `old_table_name != new_table_name`
2. `None` is equal to itself -- so if `db_table` is never specified, you get an `AlterModelTable` which changes the `db_table` to `None`.

Before fixing this, please fix the tests so that a problem like this gets caught.

Reply:
The following changes should fix this without the need to change the tests. 🤔 
```diff
index e5f2d53f45..0b60ae60fb 100644
--- a/django/db/migrations/autodetector.py
+++ b/django/db/migrations/autodetector.py
@@ -610,11 +610,11 @@ class MigrationAutodetector:
                 ):
                     continue
                 dependencies = []
-                old_table_name = rem_model_state.options.get("db_table")
-                new_table_name = model_state.options.get("db_table")
+                old_table_name = rem_model_state.options.get("db_table", f"{rem_app_label}_{rem_model_name}")
+                new_table_name = model_state.options.get("db_table", f"{app_label}_{model_name}")
                 # If moved model has same table name as it was in old app,
                 # then don't rename table.
-                if old_table_name == new_table_name:
+                if old_table_name != new_table_name:
                     self.add_operation(
```

Reply:
One of the main purposes of the tests is to tell us when the code is broken. When we find out that the code could be broken without causing the tests to fail, that is reason to fix the tests -- so at least this specific bug will not be repeated.

IMO the tests need to include at least three cases: 
- `db_table` not set at all;
- `db_table` set and remains the same after the move;
- `db_table` set and changed after the move.

Other interesting cases, which I'm not sure are mandatory, are:
- `db_table` not set before the move, but set to the old default after the move; and
- `db_table` set (to the new default or otherwise) before the move, and not set after the move. 

Checking the details of the generated operations could also have caught this, I don't think it's a good replacement for checking the different cases.

The suggested change should handle all the cases correctly, AFAICT.

Reply:
Thanks 👍 ,I did test all the cases locally before suggesting changes here and all were working correctly.🥳 
Also, I have added the tests for initial three cases as you suggested.